src/utils/viewer/viewer.cpp

```diff
@@ -769,6 +791,48 @@ void Viewer::drawEvent() {
   redraw();
 }
 
+void Viewer::moveAndLook(int repetitions) {
+  for (int i = 0; i < repetitions; i++) {
+    if (keysPressed[KeyEvent::Key::Left]) {
+      defaultAgent_->act("turnLeft");
+    }
+    if (keysPressed[KeyEvent::Key::Right]) {
+      defaultAgent_->act("turnRight");
+    }
+    if (keysPressed[KeyEvent::Key::Up]) {
+      defaultAgent_->act("lookUp");
+    }
+    if (keysPressed[KeyEvent::Key::Down]) {
+      defaultAgent_->act("lookDown");
+    }
+
+    if (keysPressed[KeyEvent::Key::A]) {
+      defaultAgent_->act("moveLeft");
+      recAgentLocation();
+    }
+    if (keysPressed[KeyEvent::Key::D]) {
+      defaultAgent_->act("moveRight");
+      recAgentLocation();
+    }
+    if (keysPressed[KeyEvent::Key::S]) {
+      defaultAgent_->act("moveBackward");
+      recAgentLocation();
+    }
+    if (keysPressed[KeyEvent::Key::W]) {
+      defaultAgent_->act("moveForward");
+      recAgentLocation();
+    }
+    if (keysPressed[KeyEvent::Key::X]) {
+      defaultAgent_->act("moveDown");
+      recAgentLocation();
+    }
+    if (keysPressed[KeyEvent::Key::Z]) {
+      defaultAgent_->act("moveUp");
+      recAgentLocation();
+    }
```

Comment:
This is very close, but you don't really want to record the movements quite like this - we don't want to record multiple locations in a single pass, considering these changes together are all reflecting a desired single movement.  We'd prefer this to be a kind of latch, where any movement would set a flag that would then prompt the recAgentLocation method.

In other words, set a boolean to true if any of the changes happen, and if the bool is true then call recAgentLocation() 1 time.

So, at line 808, something like 
`bool moved = false;`

and then in every if check : 
`moved = true;`

and finally, after the Z check is completed (line 833) : 

```
if moved {
     recAgentLocation();
}
```

Reply:
Ah, got it--sorry about that. Just checking: if the agent is moved several times in a single frame due to lag (i.e. repetitions is greater than 1), should we record the agent's location at each step between the frames, or only once per frame? I assumed the former, which I believe corresponds to calling recAgentLocation inside the for loop, and just pushed a commit with this.

Reply:
Yes, that is the correct mechanism.  In viewer, these locations are being recorded so that a demonstration of the trajectory recording/traj-pipe synthesis process can be provided to the user.   We'd want to preserve every location the agent comes to rest in.